### PR TITLE
JENKINS-47376: Fix global env variables

### PR DESCRIPTION
**Pull request by MattLud**
https://issues.jenkins-ci.org/browse/JENKINS-47376

This fixes the missing global environment variables that are missing when a container runs. 

As an alternative fix, this could instead be included as part of the POD level environment variables rather than injected and I can move it there if desired.

### CODE REVIEW

**Comment by MattLud** [comment]
The alternate spot to inject these would be here -
https://github.com/jenkinsci/kubernetes-plugin/blob/4303ec3c913c80c020eb7a6880853dc1c4838325/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L124

As part of the final setup before scheduling the Pod and as part of the Pod EnvVars.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 1]
```diff
@@ -82,34 +82,35 @@
     private final String namespace;
     private final String containerName;
     private final EnvironmentExpander environmentExpander;
+    private final EnvVars globalVars;
 
-    public ContainerExecDecorator(KubernetesClient client, String podName, String containerName, String namespace, EnvironmentExpander environmentExpander) {
+    public ContainerExecDecorator(KubernetesClient client, String podName, String containerName, String namespace, EnvironmentExpander environmentExpander, EnvVars globalVars) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: can we keep binary compatibility ?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java [hunk 2]
```diff
@@ -43,7 +49,14 @@ public boolean start() throws Exception {
         client = nodeContext.connectToCloud();
 
         EnvironmentExpander env = getContext().get(EnvironmentExpander.class);
-        decorator = new ContainerExecDecorator(client, nodeContext.getPodName(), containerName, nodeContext.getNamespace(), env);
+        EnvVars globalVars = null;
+        Jenkins instance = Jenkins.getInstance();
+        DescribableList<NodeProperty<?>, NodePropertyDescriptor>globalNodeProperties = instance.getGlobalNodeProperties();
+        List<EnvironmentVariablesNodeProperty> envVarsNodePropertyList = globalNodeProperties.getAll(EnvironmentVariablesNodeProperty.class);
+        if ( envVarsNodePropertyList != null && envVarsNodePropertyList.size() != 0 ) {
+            globalVars = envVarsNodePropertyList.get(0).getEnvVars();
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java]: I assume you need to iterate through `envVarsNodePropertyList` not just take the first item

Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java]: You shouldn't unless there are multiple GlobalNodeProperties in a config.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 1 above)
Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Should be done in the latest iteration.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 3]
```diff
@@ -118,18 +123,31 @@ public Proc launch(ProcStarter starter) throws IOException {
...
-                    if (cmd.startsWith(JENKINS_HOME)) {
-                        cmdEnvs = new String[0];
-                        LOGGER.info("Skipping injection of procstarter cmdenvs due to JENKINS_HOME present");
+                List<String> procStarter =  Arrays.asList(starter.envs());
+                List<String> cmdEnvs = new  ArrayList<String>();
+                //check if the cmd is sourced from Jenkins, rather than another plugin; if so, skip cmdEnvs except for built-in ones.
+                boolean javaHome_detected = false;
+                for(String env: procStarter)
+                {
+                    if(env.contains("JAVA_HOME"))
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: I don't understand what does it mean if `JAVA_HOME` is there, what does that mean?

Comment (carlossg) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: and due to the `break` any variable after `JAVA_HOME` is ignored, so the order of vars matter

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/AbstractKubernetesPipelineTest.java [hunk 4]
```diff
@@ -96,6 +102,14 @@ public void configureCloud() throws Exception {
         JenkinsLocationConfiguration.get().setUrl(nonLocalhostUrl.toString());
 
         r.jenkins.clouds.add(cloud);
+
+        DescribableList<NodeProperty<?>, NodePropertyDescriptor> list =  r.jenkins.getGlobalNodeProperties();
+        list.getAll(hudson.slaves.EnvironmentVariablesNodeProperty.class);
```
Comment (carlossg) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/AbstractKubernetesPipelineTest.java]: unused?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 5]
```diff
@@ -126,26 +132,100 @@ public ContainerExecDecorator(KubernetesClient client, String podName, String co
...
+                List<String> cmdEnvs = new ArrayList<String>();
+                // One issue that cropped up was that when executing sh commands, we would get the jnlp agent's injected
+                // environment variables as well, causing obvious problems such as JAVA_HOME being overwritten. The
+                // unsatisfying answer was to check for the presence of JENKINS_HOME in the cmdenvs and skip if present.
+
+                // check if the cmd is sourced from Jenkins, rather than another plugin; if so, skip cmdEnvs except for
+                // built-in ones.
+                boolean javaHome_detected = false;
+                for (String env : procStarter) {
+                    if (env.contains("JAVA_HOME")) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: I don't understand what does it mean if JAVA_HOME is there?

and due to the `break` any variable after JAVA_HOME is ignored, so the order of vars matter

Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Removed that break...The JAVA_HOME comes from not being able to tell if the execProc was dispatched by another plugin, thus necessitating the entire provided environment variables to be injected or from the master, thus overwriting whatever local environment variables there were. 

At least, that the best I can understand from the various ways Jenkins envVars are provided.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/AbstractKubernetesPipelineTest.java (hunk 4 above)
Comment (MattLud) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/AbstractKubernetesPipelineTest.java]: Removed.

**Comment by pauxus** [comment]
I implemented a similiar fix (before seeing your PR, unfortunately), however, I took a slightly different approach (taking a cue from the pipeline plugin), which might be somewhat more generic.

When injecting the variables from the proc starter, I simply removed all variables from computer.environment (key AND value). So anything that is not already there must have come from another plugin (like globalVar, configFile provider, Maven Pipeline Integration etc.). Makes for shorter code and works.

The only caveat (which should be documented) is the workings of PATH+X like environment expansions, which must only be done inside container statements to work. If needed, I can adapt my changes to the current state, but this would mostly overshadow this PR.

**Comment by MattLud** [comment]
@pauxus  - Feel free to open it and rework/remove items contributed by this PR. This code makes some naive assumptions about envvars and there is probably a better means of handling proc-injected variables though this is mostly for globals. 

:+1:

**Comment by MattLud** [comment]
@pauxus  - I've opened #283 to address a related issue. Feel free to weigh in on that as well.

**Pull request merged**